tools/benchmark/src/Configuration.ts

```diff
@@ -183,15 +183,122 @@ export interface MochaExclusiveOptions {
- * validation. This means any exception thrown from either a hook or the benchmarked function will cause test
- * failure, and subsequent operations won't be run.
+ * Arguments that can be passed to `benchmark` for optional test setup/teardown.
+ * Hooks--along with the benchmarked function--are run without additional error validation.
+ * This means any exception thrown from either a hook or the benchmarked function will cause test failure,
+ * and subsequent operations won't be run.
+ * @remarks
+ *
+ * Be careful when writing non-pure benchmark functions!
+ * Benchmark.js is written with the assumption that each cycle it runs is an independent sample.
+ * This can typically be achieved by using the `onCycle` hook to reset state, with some caveats.
+ * For more details, read below.
+ *
+ * Benchmark.js runs the benchmark function in two hierarchical groups: cycles and iterations.
+ * One iteration consists of a single execution of `benchmarkFn`.
+ * Since the time taken by a single iteration might be significantly smaller than the clock resolution, benchmark
+ * dynamically decides to run a number of iterations per cycle.
+ * After a warmup period, this number is fixed across cycles (i.e. if Benchmark.js decides to run 10,000 iterations
+ * per cycle, all statistical analysis will be performed on cycles which consist of 10,000 iterations)
+ * This strategy also helps minimize noise from JITting code.
+ *
+ * Statistical analysis is performed at the cycle level: Benchmark.js treats each cycle's timing information as a data
+ * point taken from a normal distribution, and runs cycles until the root-mean error is below a threshhold or its max
+ * time has been reached.
+ * The statistical analysis it uses is invalid if cycles aren't independent trials: consider the test
+ * ```typescript
+ * const myList = [];
+ * benchmark({
+ *     title: "insert at start of a list",
+ *     benchmarkFn: () => {
+ *         myList.unshift(0);
+ *     }
+ * });
+ * ```
+ *
+ * If each cycle has 10k iterations, the first cycle will time how long it takes to repeatedly insert elements 0 through 10k
+ * into the start `myList`.
+ * The second cycle will time how long it takes to repeatedly insert elements 10k through 20k at the start, and so on.
+ * As inserting an element at the start of the list is O(list size), it's clear that cycles will take longer and longer.
+ * We can use the `onCycle` hook to alleviate this problem:
+ * ```typescript
+ * let myList = [];
+ * benchmark({
+ *     title: "insert at start of a list",
+ *     onCycle: () => {
+ *         myList = [];
+ *     }
+ *     benchmarkFn: () => {
+ *         myList.unshift(0);
+ *     }
+ * });
+ * ```
+ *
+ * With this change, it's more reasonable to model each cycle as an independent event.
+ *
+ * Note that this approach is slightly misleading in the data it measures: if Benchmark.js chooses a cycle size of 10k,
```

Comment:
this part hints at something actually mildly concerning which I don't have a way around: since benchmark.js chooses the number of iterations per cycle dynamically based on the performance of some "warmup runs" of benchmarkFn, it's possible that perf regressions for benchmarks of this sort go completely undetected. Consider the example under test here, where list insertion time takes f(n) = n nanoseconds where `n` is the size of the list. If benchmark.js decides to have a cycle of size 10k and there is no variance in the samples, it will conclude time per cycle is 10000 * 10001 / 2, so time per iteration is just around 5 microseconds.

Now say we regress list insert performance at the start to take f(n) = 2n nanoseconds, and since the function being benchmarked is slower, Benchmark.js decides to only run 5k iterations per cycle. Now, each cycle takes 2 + 4 + 6 + ... + 10000 = 5000 * 10002 / 2, so the time per iteration is.... still 5 microseconds, despite it actually being twice as slow!

This sort of thing could be fixed if we had finer-grained control over iteration count per cycle, as well as a number of other techniques.

Reply:
I think the second calculation is incorrect (the sum of the first 5000 even numbers should be 5000 * **5001** / 2, so the end result ends up being half of the one in the first case), but the general argument stands... the dynamic number of cycles means that different runs of the same benchmark are tricky to compare against each other, particularly if the time-per-iteration of a test can vary significantly across iterations for whatever reason (like the function not being pure). 

I'd say this is a big enough gotcha that it merits a link from the package's README to this documentation.

For reference, I'll point out that since memory usage tests don't use the Benchmark library, they don't suffer from this (I think; I hope 😄). They don't have the same cycle/iteration hierarchy nor warmup period, they just run the test function many times and use that as samples from an assumed-normal distribution; and the `beforeIteration`/`afterIteration` hooks do run for each iteration, so state can be cleared entirely between iterations. Since the (memory) measurements in this case don't have the same issue as time measurements of potentially being lower than the clock resolution, I think the approach is fine.

Reply:
the calculation is correct for the scenario I mentioned (regressing perf to f(n) = 2n), but you'd get a perceived 2x performance increase with no change in code if benchmark.js chose a cycle length that's half as big and that would align with the 5000 * 5001 / 2 scenario. there's in general just a lot of issues in the space. I opened #14309 for this.